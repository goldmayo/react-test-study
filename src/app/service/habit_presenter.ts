import React from "react";
import { v4 as uuidv4 } from "uuid";

export type HabitElement = { id: string; name: string; count: number };

export interface HabitPresenterInterface {
  getHabits(): HabitElement[];
  increment(habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  decrement(habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  delete(habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  add(name: string, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  reset(update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
}

export class HabitPresenter implements HabitPresenterInterface {
  private habits: HabitElement[];
  private MAX_HABITS: number;

  constructor(habits: HabitElement[], maxHabits: number = 10) {
    this.habits = habits;
    this.MAX_HABITS = maxHabits;
  }

  getHabits = () => {
    return this.habits;
  };

  increment = (habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.map((item) => {
      if (item.id === habit.id) {
        return { ...habit, count: item.count + 1 };
      }
      return item;
    });
    update(this.habits);
  };

  decrement = (habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.map((item) => {
      if (item.id === habit.id) {
        const count = item.count - 1;
        return { ...habit, count: count < 0 ? 0 : count };
      }
      return item;
    });
    update(this.habits);
  };

  delete = (habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.filter((item) => item.id !== habit.id);
    update(this.habits);
  };

  add = (name: string, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    if (this.habits.length === this.MAX_HABITS) {
      throw new Error(`number of habits does not over the ${this.MAX_HABITS}`);
    }
    this.habits = [...this.habits, { id: uuidv4(), name: name, count: 0 }];
    update(this.habits);
  };

  reset = (update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.map((habit) => {
      if (habit.count !== 0) {
        return { ...habit, count: 0 };
      }
      return habit;
    });
    update(this.habits);
  };
}
