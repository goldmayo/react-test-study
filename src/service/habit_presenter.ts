import React from "react";
import { v4 as uuidv4 } from "uuid";

export interface HabitPresenterInterface {
  getHabits(): HabitElement[];
  increment(habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  handleDecrement(habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  handleDelete(habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  handleAdd(name: string, update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
  handleReset(update: React.Dispatch<React.SetStateAction<HabitElement[]>>): void;
}

export type HabitElement = { id: string; name: string; count: number };

// Singleton
export class HabitPresenter implements HabitPresenterInterface {
  private static instance: HabitPresenter;

  private habits: HabitElement[];

  private constructor(habits: HabitElement[]) {
    this.habits = habits;
  }

  public static getInstance(habits: HabitElement[]): HabitPresenter {
    if (!HabitPresenter.instance) {
      HabitPresenter.instance = new HabitPresenter(habits);
    }
    return HabitPresenter.instance;
  }

  getHabits = () => {
    return this.habits;
  };

  increment = (habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.map((item) => {
      if (item.id === habit.id) {
        return { ...habit, count: habit.count + 1 };
      }
      return item;
    });
    update(this.habits);
  };

  handleDecrement = (habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.map((item) => {
      if (item.id === habit.id) {
        const count = habit.count - 1;
        return { ...habit, count: count < 0 ? 0 : count };
      }
      return item;
    });
    update(this.habits);
  };

  handleDelete = (habit: HabitElement, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.filter((item) => item.id !== habit.id);
    update(this.habits);
  };

  handleAdd = (name: string, update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = [...this.habits, { id: uuidv4(), name: name, count: 0 }];
    update(this.habits);
  };

  handleReset = (update: React.Dispatch<React.SetStateAction<HabitElement[]>>) => {
    this.habits = this.habits.map((habit) => {
      if (habit.count !== 0) {
        return { ...habit, count: 0 };
      }
      return habit;
    });
    update(this.habits);
  };
}
