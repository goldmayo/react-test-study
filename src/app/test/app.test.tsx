import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TestRenderer from "react-test-renderer";

import App from "../app";
import { HabitPresenter } from "../service/habit_presenter";

describe("App", () => {
  let app: React.ReactElement;
  let presenter: HabitPresenter;
  beforeEach(() => {
    presenter = new HabitPresenter(
      [
        { id: "1", name: "Reading", count: 0 },
        { id: "2", name: "Eating", count: 0 },
        { id: "3", name: "studying", count: 1 },
      ],
      5
    );
    app = <App presenter={presenter} />;
  });
  it("render", () => {
    const component = TestRenderer.create(app);
    expect(component.toJSON()).toMatchSnapshot();
  });
  describe("App Component", () => {
    it("counts only active habits", () => {
      render(app);
      const button: HTMLElement = screen.getAllByTitle("increment")[0];
      userEvent.click(button);
      const count = screen.getByTitle("total-count");
      expect(count.innerHTML).toBe("2");
    });
    it("add new habit", () => {
      render(app);
      const newHabit = "New Habit";
      const input: HTMLElement = screen.getByPlaceholderText("please enter your habit");
      const button: HTMLElement = screen.getByText("Add");
      userEvent.type(input, newHabit);
      userEvent.click(button);
      const addedName = screen.getAllByTitle("habit-name")[0];
      expect(addedName.innerHTML).toBe("Reading");
      const addedCount = screen.getAllByTitle("habit-count")[0];
      expect(addedCount.innerHTML).toBe("0");
    });
    it("delete an item", () => {
      render(app);
      const firstHabitButton: HTMLElement = screen.getAllByTitle("delete")[0];
      userEvent.click(firstHabitButton);
      const nextHabitName = screen.getAllByTitle("habit-name")[0];
      expect(nextHabitName.innerHTML).not.toBe("Reading");
    });
    it("increase the counter", () => {
      render(app);
      const button: HTMLElement = screen.getAllByTitle("increment")[0];
      userEvent.click(button);
      const count = screen.getAllByTitle("habit-count")[0];
      expect(count.innerHTML).toBe("1");
    });
    it("decrease the counter", () => {
      render(app);
      const button: HTMLElement = screen.getAllByTitle("decrement")[2];
      userEvent.click(button);
      const count = screen.getAllByTitle("habit-count")[2];
      expect(count.innerHTML).toBe("0");
    });
    it("resets all counters", () => {
      render(app);
      const button: HTMLElement = screen.getByText("ResetAll");
      userEvent.click(button);
      screen.getAllByTitle("habit-count").forEach((count) => {
        expect(count.innerHTML).toBe("0");
      });
    });
  });
});
