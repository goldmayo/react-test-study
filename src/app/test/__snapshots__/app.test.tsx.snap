// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render 1`] = `
Array [
  <header>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-leaf header-logo"
      data-icon="leaf"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M512 165.4c0 127.9-70.05 235.3-175.3 270.1c-20.04 7.938-41.83 12.46-64.69 12.46c-64.9 0-125.2-36.51-155.7-94.47c-54.13 49.93-68.71 107-68.96 108.1C44.72 472.6 34.87 480 24.02 480c-1.844 0-3.727-.2187-5.602-.6562c-12.89-3.098-20.84-16.08-17.75-28.96c9.598-39.5 90.47-226.4 335.3-226.4C344.8 224 352 216.8 352 208S344.8 192 336 192C228.6 192 151 226.6 96.29 267.6c.1934-10.82 1.242-21.84 3.535-33.05c13.47-65.81 66.04-119 131.4-134.2c28.33-6.562 55.68-6.013 80.93-.0054c56 13.32 118.2-7.412 149.3-61.24c5.664-9.828 20.02-9.516 24.66 .8282C502.7 76.76 512 121.9 512 165.4z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    <span>
      Habit Tracker
    </span>
    <span
      className="header-count"
      title="total-count"
    >
      1
    </span>
  </header>,
  <form
    className="add-form"
    onSubmit={[Function]}
  >
    <input
      className="add-input"
      placeholder="please enter your habit"
      type="text"
    />
    <button
      className="add-button"
    >
      Add
    </button>
  </form>,
  <ul>
    <li
      className="habit"
    >
      <span
        className="habit-name"
        title="habit-name"
      >
        Reading
      </span>
      <span
        className="habit-count"
        title="habit-count"
      >
        0
      </span>
      <button
        className="habit-button habit-increase"
        onClick={[Function]}
        title="increment"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-plus "
          data-icon="square-plus"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M200 344V280H136C122.7 280 112 269.3 112 256C112 242.7 122.7 232 136 232H200V168C200 154.7 210.7 144 224 144C237.3 144 248 154.7 248 168V232H312C325.3 232 336 242.7 336 256C336 269.3 325.3 280 312 280H248V344C248 357.3 237.3 368 224 368C210.7 368 200 357.3 200 344zM0 96C0 60.65 28.65 32 64 32H384C419.3 32 448 60.65 448 96V416C448 451.3 419.3 480 384 480H64C28.65 480 0 451.3 0 416V96zM48 96V416C48 424.8 55.16 432 64 432H384C392.8 432 400 424.8 400 416V96C400 87.16 392.8 80 384 80H64C55.16 80 48 87.16 48 96z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="habit-button habit-decrease"
        onClick={[Function]}
        title="decrement"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-minus "
          data-icon="square-minus"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M312 232C325.3 232 336 242.7 336 256C336 269.3 325.3 280 312 280H136C122.7 280 112 269.3 112 256C112 242.7 122.7 232 136 232H312zM0 96C0 60.65 28.65 32 64 32H384C419.3 32 448 60.65 448 96V416C448 451.3 419.3 480 384 480H64C28.65 480 0 451.3 0 416V96zM48 96V416C48 424.8 55.16 432 64 432H384C392.8 432 400 424.8 400 416V96C400 87.16 392.8 80 384 80H64C55.16 80 48 87.16 48 96z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="habit-button habit-delete"
        onClick={[Function]}
        title="delete"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-trash "
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.69C140.6 6.848 151.7 0 163.8 0H284.2C296.3 0 307.4 6.848 312.8 17.69L320 32H416C433.7 32 448 46.33 448 64C448 81.67 433.7 96 416 96H32C14.33 96 0 81.67 0 64C0 46.33 14.33 32 32 32H128L135.2 17.69zM394.8 466.1C393.2 492.3 372.3 512 346.9 512H101.1C75.75 512 54.77 492.3 53.19 466.1L31.1 128H416L394.8 466.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </li>
    <li
      className="habit"
    >
      <span
        className="habit-name"
        title="habit-name"
      >
        Eating
      </span>
      <span
        className="habit-count"
        title="habit-count"
      >
        0
      </span>
      <button
        className="habit-button habit-increase"
        onClick={[Function]}
        title="increment"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-plus "
          data-icon="square-plus"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M200 344V280H136C122.7 280 112 269.3 112 256C112 242.7 122.7 232 136 232H200V168C200 154.7 210.7 144 224 144C237.3 144 248 154.7 248 168V232H312C325.3 232 336 242.7 336 256C336 269.3 325.3 280 312 280H248V344C248 357.3 237.3 368 224 368C210.7 368 200 357.3 200 344zM0 96C0 60.65 28.65 32 64 32H384C419.3 32 448 60.65 448 96V416C448 451.3 419.3 480 384 480H64C28.65 480 0 451.3 0 416V96zM48 96V416C48 424.8 55.16 432 64 432H384C392.8 432 400 424.8 400 416V96C400 87.16 392.8 80 384 80H64C55.16 80 48 87.16 48 96z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="habit-button habit-decrease"
        onClick={[Function]}
        title="decrement"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-minus "
          data-icon="square-minus"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M312 232C325.3 232 336 242.7 336 256C336 269.3 325.3 280 312 280H136C122.7 280 112 269.3 112 256C112 242.7 122.7 232 136 232H312zM0 96C0 60.65 28.65 32 64 32H384C419.3 32 448 60.65 448 96V416C448 451.3 419.3 480 384 480H64C28.65 480 0 451.3 0 416V96zM48 96V416C48 424.8 55.16 432 64 432H384C392.8 432 400 424.8 400 416V96C400 87.16 392.8 80 384 80H64C55.16 80 48 87.16 48 96z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="habit-button habit-delete"
        onClick={[Function]}
        title="delete"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-trash "
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.69C140.6 6.848 151.7 0 163.8 0H284.2C296.3 0 307.4 6.848 312.8 17.69L320 32H416C433.7 32 448 46.33 448 64C448 81.67 433.7 96 416 96H32C14.33 96 0 81.67 0 64C0 46.33 14.33 32 32 32H128L135.2 17.69zM394.8 466.1C393.2 492.3 372.3 512 346.9 512H101.1C75.75 512 54.77 492.3 53.19 466.1L31.1 128H416L394.8 466.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </li>
    <li
      className="habit"
    >
      <span
        className="habit-name"
        title="habit-name"
      >
        studying
      </span>
      <span
        className="habit-count"
        title="habit-count"
      >
        1
      </span>
      <button
        className="habit-button habit-increase"
        onClick={[Function]}
        title="increment"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-plus "
          data-icon="square-plus"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M200 344V280H136C122.7 280 112 269.3 112 256C112 242.7 122.7 232 136 232H200V168C200 154.7 210.7 144 224 144C237.3 144 248 154.7 248 168V232H312C325.3 232 336 242.7 336 256C336 269.3 325.3 280 312 280H248V344C248 357.3 237.3 368 224 368C210.7 368 200 357.3 200 344zM0 96C0 60.65 28.65 32 64 32H384C419.3 32 448 60.65 448 96V416C448 451.3 419.3 480 384 480H64C28.65 480 0 451.3 0 416V96zM48 96V416C48 424.8 55.16 432 64 432H384C392.8 432 400 424.8 400 416V96C400 87.16 392.8 80 384 80H64C55.16 80 48 87.16 48 96z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="habit-button habit-decrease"
        onClick={[Function]}
        title="decrement"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-minus "
          data-icon="square-minus"
          data-prefix="far"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M312 232C325.3 232 336 242.7 336 256C336 269.3 325.3 280 312 280H136C122.7 280 112 269.3 112 256C112 242.7 122.7 232 136 232H312zM0 96C0 60.65 28.65 32 64 32H384C419.3 32 448 60.65 448 96V416C448 451.3 419.3 480 384 480H64C28.65 480 0 451.3 0 416V96zM48 96V416C48 424.8 55.16 432 64 432H384C392.8 432 400 424.8 400 416V96C400 87.16 392.8 80 384 80H64C55.16 80 48 87.16 48 96z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="habit-button habit-delete"
        onClick={[Function]}
        title="delete"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-trash "
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.69C140.6 6.848 151.7 0 163.8 0H284.2C296.3 0 307.4 6.848 312.8 17.69L320 32H416C433.7 32 448 46.33 448 64C448 81.67 433.7 96 416 96H32C14.33 96 0 81.67 0 64C0 46.33 14.33 32 32 32H128L135.2 17.69zM394.8 466.1C393.2 492.3 372.3 512 346.9 512H101.1C75.75 512 54.77 492.3 53.19 466.1L31.1 128H416L394.8 466.1z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </li>
  </ul>,
  <button
    className="habit-reset"
    onClick={[Function]}
  >
    ResetAll
  </button>,
]
`;
